Optimize source refresh and ES indexation process

LePetitTim opened the pull request.

LePetitTim committed:
⚡️ Change elastic search indexation on feature + savepoint commit insertion

LePetitTim committed:
🐛 fix build front already exist

LePetitTim committed:
🐛 Failure state no model should not broke

LePetitTim committed:
✅ Add tests / fix mocks ES

LePetitTim committed:
⚡️ Use manual savepoints, add report exception

LePetitTim committed:
⚡️ Improve performance refresh data

LePetitTim committed:
✅ Add tests exception

Commit:
Rework source reporting & move it to a model

Commit:
✨ Add modified lines count in source reporting

Commit:
✨ Fix wrong deleted lines report in source reporting

Commit:
🚨 Code style: Black & isort

Commit:
✅ Add tests for source reporting

submarcos commented on project/geosource/models.py:
```diff
@@ -63,6 +63,39 @@ def get_type_from_data(cls, data):
+    errors = models.JSONField(default=list)
+
+    def save(self, *args, **kwargs):
+        self.ended = timezone.now()
```
I think that is task report method that should update this field

CFarcy replied:
Are you talking about the methods in `project/geosource.task.py` ?

CFarcy replied:
Done !

submarcos commented on project/geosource/models.py:
```diff
@@ -76,7 +109,7 @@ class Source(PolymorphicModel, CeleryCallMethodsMixin):
     settings = models.JSONField(default=dict, blank=True)
     groups = models.ManyToManyField(Group, blank=True, related_name="geosources")
-    report = models.JSONField(default=dict, blank=True, encoder=DjangoJSONEncoder)
+    report = models.ForeignKey(SourceReporting, on_delete=models.PROTECT, null=True)
```
report i nullable, so deleting report should set NULL on delete

CFarcy replied:
Yes correct, I put protected when I start the dev and forgot to change it.

CFarcy replied:
Done !

CFarcy commented on project/geosource/models.py:
```diff
@@ -63,6 +63,39 @@ def get_type_from_data(cls, data):
+class SourceReporting(models.Model):
+    SUCCESS = "0"
+    ERROR = "1"
+    WARNING = "2"
```
Oh, nice !

CFarcy replied:
Done !

Commit:
Use textchoice for source reporting status

Commit:
Move ended datetime of source reporting to tasks

Commit:
✏️ fix type in SourceReporting choices

Commit:
Change report foreignkey as oneToOne

Commit:
Add pending status to source reporting

Commit:
Add more ES mock to make tests pass correctly

Commit:
Filling coverage

Commit:
More coverage

Commit:
Add generic status to Source

Commit:
Update test

Commit:
Refactoring source reporting

- Updating & saving the SourceReport instance only in refresh_data method
 - Use exception in lower methods to handle errors & to raise them to upper method (refresh_data)

submarcos commented on project/geosource/models.py:
```diff
@@ -63,7 +63,54 @@ def get_type_from_data(cls, data):
+class CSVSourceException(SourceException):
+    """CSVSource exception raised by the CSVSource model"""
+
+    pass
```
can you move these in dedicated file ?

CFarcy replied:
Yes, absolutely!

submarcos commented on project/geosource/models.py:
```diff
@@ -76,13 +123,16 @@ class Source(PolymorphicModel, CeleryCallMethodsMixin):
     updated_at = models.DateTimeField(auto_now=True)
     last_refresh = models.DateTimeField(default=timezone.now)
+    status = models.PositiveIntegerField(
+        choices=Status.choices, null=True, max_length=20
```
```suggestion
        choices=Status.choices, null=True
```
no need to get a max_length on an integer. Bu you can use PositiveSmallIntegerField

CFarcy replied:
Oh, that's nice. I will definitively use `PositiveSmallIntegerField`

CFarcy commented on project/geosource/models.py:
```diff
@@ -63,7 +63,54 @@ def get_type_from_data(cls, data):
+        SUCCESS = 0, "Success"
+        ERROR = 1, "Error"
+        WARNING = 2, "Warning"
+        PENDING = 3, "Pending"
```
Yes, you're right.  I think the report status (and even errors messages) could be translated values also !

Commit:
Update project/geosource/models.py

Co-authored-by: Jean-Etienne Castagnede <j.e.castagnede@gmail.com>

Commit:
Update project/geosource/models.py

Co-authored-by: Jean-Etienne Castagnede <j.e.castagnede@gmail.com>

Commit:
Update project/geosource/serializers.py

Co-authored-by: Jean-Etienne Castagnede <j.e.castagnede@gmail.com>

Commit:
Update project/geosource/models.py

Co-authored-by: Jean-Etienne Castagnede <j.e.castagnede@gmail.com>

Commit:
Small code adjustement

Commit:
Isort

Commit:
Add translation for report status message

Commit:
Remove unused variables

Commit:
Isort

Commit:
Update migration to make it work with reel data

* Remove then "re-create" the report field on the Source model to avoid conflict with existing data

Commit:
Force report creation at refresh start

Commit:
Force save report after reset

Commit:
Uncatched exception by get_records set reporting in error state

Commit:
Report wrong user info for csvsource

Commit:
Black

Commit:
Updating tests

Commit:
Fix postgis refresh failing

Commit:
Rework source status

* Api return only status code instead of text
* wmts source does not have a refresh status

Commit:
Fix report not set to warning status correctly

Commit:
Improve source reporting

* remove unnecessary reporting
* improve message readability
* add total count to the report

Commit:
Improve test coverage

Commit:
Add ES patch to tests

Commit:
Black

Commit:
More SourceException tests

Commit:
Test refresh status for wmtssource

Commit:
Improve coverage

submarcos committed:
Update changelog.rst
